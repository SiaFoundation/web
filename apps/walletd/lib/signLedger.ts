import { SiacoinElement, Transaction } from '@siafoundation/react-walletd'
import { getWalletWasm } from './wasm'
import { stripPrefix } from '@siafoundation/design-system'
import { AddressData } from '../contexts/addresses/types'
import { LedgerDevice } from '../contexts/ledger/types'
import { addUnlockConditionsAndSignatures, getUtxoAndAddress } from './sign'

export async function signTransactionLedger({
  device,
  transaction,
  toSign,
  addresses,
  siacoinOutputs,
}: {
  device: LedgerDevice
  transaction: Transaction
  toSign: string[]
  addresses: AddressData[]
  siacoinOutputs: SiacoinElement[]
}): Promise<{ transaction?: Transaction; error?: string }> {
  if (!addresses) {
    return { error: 'No addresses' }
  }
  if (!siacoinOutputs) {
    return { error: 'No outputs' }
  }

  const { error } = addUnlockConditionsAndSignatures({
    transaction,
    toSign,
    addresses,
    siacoinOutputs,
  })

  if (error) {
    return { error }
  }

  // for each toSign
  for (const [i, idPrefixed] of toSign.entries()) {
    const id = stripPrefix(idPrefixed)

    // find the utxo and corresponding address
    const { address, error: utxoAddressError } = getUtxoAndAddress({
      id,
      addresses,
      siacoinOutputs,
    })
    if (utxoAddressError) {
      return { error: utxoAddressError }
    }

    // This function generates the signature and adds it to the existing transaction
    const { error } = await signTransactionIndex({
      device,
      transaction,
      signatureIndex: i,
      keyIndex: address.index,
    })

    if (error) {
      return {
        error,
      }
    }
  }

  return {
    transaction,
  }
}

async function signTransactionIndex({
  device,
  transaction,
  signatureIndex,
  keyIndex,
}: {
  device: LedgerDevice
  transaction: Transaction
  signatureIndex: number
  keyIndex: number
}): Promise<{ transaction?: Transaction; error?: string }> {
  const { encodedTransaction, error } = getWalletWasm().encodeTransaction(
    JSON.stringify(transaction)
  )
  if (error) {
    return { error }
  }
  const encodedTransactionBuffer = Buffer.from(encodedTransaction, 'utf-8')
  const version = await device.sia.getVersion()
  // compat: v0.4.5 introduces the change index to the sign txn ADPU
  const signCompat = versionCmp(version, '0.4.5') < 0

  let signature: string = undefined
  try {
    if (signCompat) {
      signature = await device.sia.signTransactionV044(
        encodedTransactionBuffer,
        signatureIndex,
        keyIndex
      )
    } else {
      signature = await device.sia.signTransaction(
        encodedTransactionBuffer,
        signatureIndex,
        keyIndex,
        0
      )
    }
  } catch (e) {
    return { error: e.message }
  }

  transaction.signatures[signatureIndex].signature = signature

  return {
    transaction,
  }
}

function versionCmp(a, b) {
  const reg = /[^0-9.]/gi,
    aPieces = a.replace(reg, '').split('.'),
    bPieces = b.replace(reg, '').split('.'),
    len = Math.max(aPieces.length, bPieces.length)

  for (let i = 0; i < len; i++) {
    let as = 0,
      bs = 0

    if (i < aPieces.length) as = parseInt(aPieces[i], 10)

    if (i < bPieces.length) bs = parseInt(bPieces[i], 10)

    if (isNaN(as)) as = 0

    if (isNaN(bs)) bs = 0

    if (as < bs) return -1
    else if (as > bs) return 1
  }

  return 0
}
