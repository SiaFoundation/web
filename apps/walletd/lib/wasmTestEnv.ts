import { initWasmGo } from './wasm_exec_mod'
import fs from 'fs'

export async function loadWASMTestEnv(): Promise<{ error?: string }> {
  const Go = initWasmGo()
  const go = new Go()
  const wasm = fs.readFileSync('apps/walletd/public/walletd.wasm')
  try {
    const result = await WebAssembly.instantiate(wasm, go.importObject)
    go.run(result.instance)
    return {}
  } catch (e) {
    console.log(e)
    return {
      error: e.message,
    }
  }
}
