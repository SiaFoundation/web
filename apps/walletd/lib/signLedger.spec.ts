import { signTransactionLedger } from './signLedger'
import { TextEncoder, TextDecoder } from 'util'
import { loadWASMTestEnv } from './wasmTestEnv'
import {
  getMockDevice,
  getAddresses,
  getSiacoinOutputs,
  getSiafundOutputs,
  getToSign,
  getTransaction,
} from './testMocks'
global.TextEncoder = TextEncoder
global.TextDecoder = TextDecoder

describe('signLedger', () => {
  describe('siacoin', () => {
    it('builds and signs valid transaction', async () => {
      await loadWASMTestEnv()
      const device = getMockDevice()
      expect(
        await signTransactionLedger({
          device,
          transaction: getTransaction(),
          toSign: getToSign(),
          addresses: getAddresses(),
          siacoinOutputs: getSiacoinOutputs(),
          siafundOutputs: getSiafundOutputs(),
        })
      ).toMatchSnapshot()
    })

    it('errors when a toSign utxo is missing', async () => {
      await loadWASMTestEnv()
      const device = getMockDevice()
      expect(
        await signTransactionLedger({
          device,
          transaction: getTransaction(),
          toSign: [getToSign()[0], 'not in siacoinOutputs'],
          addresses: getAddresses(),
          siacoinOutputs: getSiacoinOutputs(),
          siafundOutputs: getSiafundOutputs(),
        })
      ).toEqual({
        error: 'Missing utxo',
      })
    })

    it('errors when a public keys address is missing', async () => {
      await loadWASMTestEnv()
      const device = getMockDevice()
      expect(
        await signTransactionLedger({
          device,
          transaction: getTransaction(),
          toSign: getToSign(),
          addresses: [
            {
              id: 'id',
              walletId: 'id',
              address: 'address not in addresses',
              index: 5,
            },
          ],
          siacoinOutputs: getSiacoinOutputs(),
          siafundOutputs: getSiafundOutputs(),
        })
      ).toEqual({
        error: 'Missing address',
      })
    })

    it('errors when an address is missing its index', async () => {
      await loadWASMTestEnv()
      const device = getMockDevice()
      expect(
        await signTransactionLedger({
          device,
          transaction: getTransaction(),
          toSign: getToSign(),
          addresses: [
            {
              id: 'id',
              walletId: 'id',
              address: getSiacoinOutputs()[1].siacoinOutput.address,
            },
          ],
          siacoinOutputs: getSiacoinOutputs(),
          siafundOutputs: getSiafundOutputs(),
        })
      ).toEqual({
        error: 'Missing address index',
      })
    })

    it('errors when an address is missing its public key', async () => {
      await loadWASMTestEnv()
      const device = getMockDevice()
      const addresses = getAddresses()
      addresses[0].publicKey = undefined
      expect(
        await signTransactionLedger({
          device,
          transaction: getTransaction(),
          toSign: getToSign(),
          addresses,
          siacoinOutputs: getSiacoinOutputs(),
          siafundOutputs: getSiafundOutputs(),
        })
      ).toEqual({
        error: 'Missing address public key',
      })
    })
  })
})
