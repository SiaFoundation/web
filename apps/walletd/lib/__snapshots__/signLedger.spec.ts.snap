// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`signLedger siacoin builds and signs valid transaction 1`] = `
Object {
  "transaction": Object {
    "minerFees": Array [
      "3930000000000000000000",
    ],
    "siacoinInputs": Array [
      Object {
        "parentID": "scoid:b222428602c8382b67a769d17e1cdc0952f37f2441a872b92671a6ed76cf22f5",
        "unlockConditions": Object {
          "publicKeys": Array [
            "ed25519:b5b9196a3c19f94982bcdba250a973181b22112437832a8f818f4aa73b8add74",
          ],
          "signaturesRequired": 1,
          "timelock": 0,
        },
      },
    ],
    "siacoinOutputs": Array [
      Object {
        "address": "addr:934b885229a9f98153401d7a647a1862aede399c656f33ec8492dfffca557ca907a3d22089c8",
        "value": "95408980544305197274920800",
      },
    ],
    "siafundInputs": Array [
      Object {
        "claimAddress": "addr:934b885229a9f98153401d7a647a1862aede399c656f33ec8492dfffca557ca907a3d22089c8",
        "parentID": "sfoid:b53e88ce69f19f0bf1d3496479f20b72e1133c719e82278830ee6618bb582852",
        "unlockConditions": Object {
          "publicKeys": Array [
            "ed25519:8a7496aa59f17a4aae68c7e41e09d5ca94e64ba27f74cdb0b143f70dcc67b206",
          ],
          "signaturesRequired": 1,
          "timelock": 0,
        },
      },
    ],
    "siafundOutputs": Array [
      Object {
        "address": "addr:eb2ee5169dd9aaab804b38f7e70043690ac21da1144990a4a28c1dcf66cd7ee9845aef03006f",
        "value": 1,
      },
    ],
    "signatures": Array [
      Object {
        "coveredFields": Object {
          "wholeTransaction": true,
        },
        "parentID": "b53e88ce69f19f0bf1d3496479f20b72e1133c719e82278830ee6618bb582852",
        "publicKeyIndex": 0,
        "signature": "Xt1EJckLmWXU+7HHHDN9bRV5KRuLdC4YY01LzaAMF269QH4hWV8zFkY3kCWs65svhb9HhA1Ix1MRGvhN9orBDpAA",
        "timelock": 0,
      },
      Object {
        "coveredFields": Object {
          "wholeTransaction": true,
        },
        "parentID": "b222428602c8382b67a769d17e1cdc0952f37f2441a872b92671a6ed76cf22f5",
        "publicKeyIndex": 0,
        "signature": "fvmSaRzlO/n2L5tsT32e82kWqHnIjQJ8cqjWOc37TtlK6p/vIiOG+TO98HfvbgObTOYVqlKMtUyxTOjGb3bfCpAA",
        "timelock": 0,
      },
    ],
  },
}
`;
